### PR TITLE
Cloud function for syncing OSS-Fuzz projects (#3538).

### PR DESCRIPTION
**Pull request by kabeer27**
Integrating the first cloud function i implemented which syncs the project list from github and uploads the list to cloud datastore, which will be used by another cloud function to request builds.

@oliverchang 
@inferno-chromium

### CODE REVIEW

File: infra/build/functions/main.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+#!/bin/bash -eu
```
Comment (oliverchang) [inline infra/build/functions/main.py:1]: This is incorrect, remove.

File: infra/build/functions/main.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      project.name for project in Project.query()
+      if project.name in projects
+    }
+
+  for project in projects:
+    if project not in already_exist:
+      Project(name=project).put()
+
+def _has_docker_file(repo, project_path):
+  return any(content_file.name == "Dockerfile"
```
Comment (oliverchang) [inline infra/build/functions/main.py:53]: Use single quotes for all strings (not including docstrings).

File: infra/build/functions/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+PyGithub
```
Comment (oliverchang) [inline infra/build/functions/requirements.txt:18]: Version these.

File: infra/build/functions/test_sync.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
+#!/bin/bash -eu
```
Comment (oliverchang) [inline infra/build/functions/test_sync.py:1]: Again, the bash shebang is wrong. Remove.

File: infra/build/functions/main.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+def sync_projects(projects):
+  """Sync projects with cloud datastore"""
+  to_remove = [
+      project for project in Project.query()
+      if project.name not in projects
+  ]
+
+  for project in to_remove:
+    project.key.delete()
```
Comment (oliverchang) [inline infra/build/functions/main.py:41]: you can just use ndb.delete_multi(to_remove) here. You need to update the above to make `to_remove` contain keys rather than the actual entities.

File: infra/build/functions/main.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+  already_exist = {
+      project.name for project in Project.query()
+      if project.name in projects
+    }
```
Comment (oliverchang) [inline infra/build/functions/main.py:46]: Please run infra/presubmit.py {lint,format} and fix all lint/formatting issues.

File: infra/build/functions/main.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  """Sync projects with cloud datastore"""
+  to_remove = [
+      project for project in Project.query()
+      if project.name not in projects
+  ]
+
+  for project in to_remove:
+    project.key.delete()
+
+  already_exist = {
```
Comment (oliverchang) [inline infra/build/functions/main.py:43]: rename to `existing_projects`.

File: infra/build/functions/main.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+      and VALID_PROJECT_NAME.match(content_file.name)
+  }
+  return projects
+
+def sync(event, context):
+  """sync with cloud datastore"""
+  client = ndb.Client()
+
+  with client.context():
+    token = iter(GitAuth.query()).next()
```
Comment (oliverchang) [inline infra/build/functions/main.py:72]: just do GitAuth.query.get(). Also check for None and raise an RuntimeError exception in that case.

File: infra/build/functions/main.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  to_remove = [
+      project for project in Project.query()
+      if project.name not in projects
+  ]
+
+  for project in to_remove:
+    project.key.delete()
+
+  already_exist = {
+      project.name for project in Project.query()
```
Comment (oliverchang) [inline infra/build/functions/main.py:44]: Rather than running multiple Project.query(), we can just save it and do it once.

File: infra/build/functions/main.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+      if project.name in projects
```
Comment (oliverchang) [inline infra/build/functions/main.py:45]: Why do we need this check at all?

File: infra/build/functions/main.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+  """Datastore Entity Project"""
+  name = ndb.StringProperty()
+
+class GitAuth(ndb.Model):
+  """Datastore Entity GitAuth"""
+  access_token = ndb.StringProperty()
+
+def sync_projects(projects):
+  """Sync projects with cloud datastore"""
+  to_remove = [
```
Comment (oliverchang) [inline infra/build/functions/main.py:35]: rename to `projects_to_remove`

File: infra/build/functions/test_sync.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+  thread.daemon = True
+  thread.start()
+  if not ready_event.wait(timeout):
+    raise RuntimeError(
+        '{} emulator did not get ready in time.'.format(emulator))
+  return thread
+
+
+# pylint: disable=too-few-public-methods
+class Repository():
```
Comment (oliverchang) [inline infra/build/functions/test_sync.py:80]: no need for parentheses after the `Repository`.

File: infra/.gitignore [hunk 13]
```diff
@@ -0,0 +1,4 @@
+!build/
```
Comment (oliverchang) [inline infra/.gitignore:1]: Rather than adding this complicated .gitignore, can we just make the root level .gitignore specify an absolute build path? i.e. set `/build/` with the leading slash. 

I haven't tested this so make sure this works.

File: infra/build/functions/main.py [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+from github import Github
+from google.cloud import ndb
+
+VALID_PROJECT_NAME = re.compile(r'^[a-zA-Z0-9_-]+$')
+
+
+# pylint: disable=too-few-public-methods
+class Project(ndb.Model):
+  """Datastore Entity Project"""
```
Comment (oliverchang) [inline infra/build/functions/main.py:28]: Please give more descriptive docstrings.

I.e. "Represents an integrated OSS-Fuzz project". Same for all other classes you define.

File: infra/build/functions/main.py [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+      Project(name=project).put()
+
+
+def _has_docker_file(repo, project_path):
+  return any(content_file.name == 'Dockerfile'
+             for content_file in repo.get_contents(project_path))
+
+
+def get_projects(repo):
+  """get projects from git repo"""
```
Comment (oliverchang) [inline infra/build/functions/main.py:60]: Please capitalize the first word and add a period ('.') to the sentence. Same with all other docstrings.

File: infra/build/functions/main.py [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+  ndb.delete_multi(projects_to_remove)
+
+  existing_projects = {project.name for project in project_query}
+
+  for project in projects:
+    if project not in existing_projects:
+      Project(name=project).put()
+
+
+def _has_docker_file(repo, project_path):
```
Comment (oliverchang) [inline infra/build/functions/main.py:54]: Add a docstring here.

File: infra/build/functions/main.py [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+def sync_projects(projects):
+  """Sync projects with cloud datastore"""
+  project_query = Project.query()
+  projects_to_remove = [
+      project.key for project in project_query if project.name not in projects
+  ]
+
+  ndb.delete_multi(projects_to_remove)
+
+  existing_projects = {project.name for project in project_query}
```
Comment (oliverchang) [inline infra/build/functions/main.py:47]: Have you verified that we can iterate over the same query more than once?

File: infra/build/functions/main.py [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+
+  for project in projects:
```
Comment (oliverchang) [inline infra/build/functions/main.py:49]: nit: do this instead. 

```python
new_projects = [project for project in projects if project not in existing_projects]
ndb.put_multi(new_projects)
```

File: infra/build/functions/main.py [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+        raise RuntimeError("No access token provided")
+    except RuntimeError as re:
+      print(re)
+    access_token = token.access_token
+
+  github_client = Github(access_token)
+  repo = github_client.get_repo('google/oss-fuzz')
+  projects = get_projects(repo)
+
+  with client.context():
```
Comment (oliverchang) [inline infra/build/functions/main.py:89]: Why not use the same context from before?

File: infra/build/functions/main.py [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+def sync(event, context):
+  """sync with cloud datastore"""
+  client = ndb.Client()
+
+  with client.context():
+    try:
+      token = GitAuth.query().get()
+      if token is None:
+        raise RuntimeError("No access token provided")
+    except RuntimeError as re:
```
Comment (oliverchang) [inline infra/build/functions/main.py:81]: There's no point in the exception if you're just going to handle it straight away. Remove the try so that the exception can surface as part of the error reporting process.

File: infra/build/functions/main.py [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+
+def sync(event, context):
+  """sync with cloud datastore"""
+  client = ndb.Client()
+
+  with client.context():
+    try:
+      token = GitAuth.query().get()
+      if token is None:
+        raise RuntimeError("No access token provided")
```
Comment (oliverchang) [inline infra/build/functions/main.py:80]: single quotes

File: infra/build/functions/main.py (hunk 18 above)
Comment (kabeer27) [reply infra/build/functions/main.py:49]: `new_projects = [
      Project(name=project)
      for project in projects
      if project not in existing_projects
  ]
  ndb.put_multi(new_projects)`

Since ndb.put_multi require entities changing it to this

Comment (oliverchang) [reply infra/build/functions/main.py:49]: Yep, that's the right one :)

File: infra/build/functions/main.py (hunk 17 above)
Comment (kabeer27) [reply infra/build/functions/main.py:47]: I believe the unit test should be verifying this automatically since it covers 
1) An existing project -> test1
2) A project to be removed -> test2
3) A project to be added -> test3

File: infra/build/functions/test_sync.py [hunk 22]
```diff
@@ -0,0 +1,138 @@
...
+    for i in range(3):
+      name = 'test' + str(i)
+      repo.contents.append(Repository(name, 'dir', 'projects/' + name))
+      project = repo.contents[i]
+      project.contents.append(Repository('Dockerfile', 'file', 'placeholder'))
+
+    # Removing Dockerfile from project test1
+    repo.contents[1].contents.pop()
+
+    self.assertEqual(get_projects(repo), {'test0', 'test2'})
```
Comment (oliverchang) [inline infra/build/functions/test_sync.py:130]: Add separate tests for the edge cases:
- Invalid project name
- No Dockerfile
- A non-directory file in projects/

File: infra/build/functions/test_sync.py [hunk 23]
```diff
@@ -0,0 +1,138 @@
...
+
+    with client.context():
+      Project(name='test1').put()
+      Project(name='test2').put()
+
+      projects = {'test1', 'test3'}
+      sync_projects(projects)
+
+      projects_query = Project.query()
+      self.assertEqual(projects, {project.name for project in projects_query})
```
Comment (oliverchang) [inline infra/build/functions/test_sync.py:116]: Add a test for when there is no GitAuth token.

File: infra/build/functions/main.py [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+  token = GitAuth.query().get()
+  if token is None:
+    raise RuntimeError('No access token available')
+  return token.access_token
+
+
+def sync(event, context):
+  """Sync projects with cloud datastore."""
+
+  del event, context  #unused
```
Comment (oliverchang) [inline infra/build/functions/main.py:87]: Instead of doing this, please just add a `# pylint: disable=unused-argument` on line 84.
